# AcademIA - BB

# UAN

## Live 02 - Estruração de Dados

### Revisão sobre Pandas

#### Importação da biblioteca

#### Conceito de dataframe Pandas

#### Criação manual de dataframe Pandas

#### Exploração básica de dataframe

### Levantamento de participantes no programa de capacitação
- A aplicação da solução será fazer uma análise simples de participantes do AcademIA-BB. Serão carregados dados dos participantes (matricula, Nome, Dependência de Localização, Data de Ingresso, [Curso, Área de Conhecimento, Data de Conclusão, Instituição] e dados da dependência (código da UOR, nome reduzido da UOR, Código da dependência vinculado a UOR, número de dependências subordinadas).)

#### Desenvolvimento
- Para a resolução do problema, serão carregados três arquivos:
    - func_perfil.csv, 
    - local_part_01.csv
    - local_part_02.csv.
- Após carregados os arquivos referentes à UOR, eles deverão ser unidos em um só dataframe Pandas.
- Com todos os dados devidamente carregados em dataframes, estes deverão ser combinados (tanto de funcionários, como das UORs) para garantir as devidas análises.

# Importação da Biblioteca Pandas
referência: https://pandas.pydata.org/docs/.)

In [3]:
import pandas as pd

# Overview sobre dataframe

## Criação de dataframe Pandas manualmente

referências: 
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html
- https://www.w3schools.com/python/pandas/pandas_dataframes.asp
- https://www.w3schools.com/Python/python_tuples.asp
- https://www.geeksforgeeks.org/create-pyspark-dataframe-from-list-of-tuples/
- https://pandas.pydata.org/docs/user_guide/10min.html

In [4]:
# Cria a lista de registros de participantes
lista_participante = [
    ("Ana Maria", "1990-08-16", "GO", True),
    ('André Bernado', '1988-02-25', 'SP', False),
    ('Fabrícia Alencar', '1985-03-29', 'DF', False),
    ('Fabio José', '1979-02-22', None, True)
]

# criar a lista referente aos nomes da colunas do dataframe
lista_coluna = ["Nome", "Data_nascimento", "Naturalidade", "Indicador_conclusao_curso"]

# cria o dataframe
df_participante = pd.DataFrame(lista_participante, columns=lista_coluna)

### Exploração básica do dataframe

#### Mostra conteúdo do dataframe

In [5]:
df_participante.head()

Nome Data_nascimento Naturalidade  Indicador_conclusao_curso
0         Ana Maria      1990-08-16           GO                       True
1     André Bernado      1988-02-25           SP                      False
2  Fabrícia Alencar      1985-03-29           DF                      False
3        Fabio José      1979-02-22         None                       True

#### Mostra informação sobre os campos do dataframe

referência:

- https://pandas.pydata.org/docs/user_guide/10min.html

In [6]:
df_participante.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Nome                       4 non-null      object
 1   Data_nascimento            4 non-null      object
 2   Naturalidade               3 non-null      object
 3   Indicador_conclusao_curso  4 non-null      bool  
dtypes: bool(1), object(3)
memory usage: 232.0+ bytes


Mostra descrição (estatística) sobre os dados do dataframe

In [7]:
df_participante.describe()

Nome Data_nascimento Naturalidade Indicador_conclusao_curso
count           4               4            3                         4
unique          4               4            3                         2
top     Ana Maria      1990-08-16           GO                      True
freq            1               1            1                         2

# Desenvolvimento
- o desenvolvimento seguirá a ordem de um job ETL (extrac, transform, load)

## Aquisição de Dados

### Dados dos funcionários

In [8]:
import os, types
import pandas as pd

In [9]:
# no labblite

df_1 = pd.read_csv('func_perfil.csv')
df_1.head(5)

matricula              Nome  local de trabalho Data de Ingresso  \
0     236548  Sra. Laura Silva               8008       2021-06-03   
1     236549    Henrique Ramos               8008       2020-06-28   
2     236550      Marina Lopes               8008       2015-12-08   
3     236551   Dra. Laura Dias               8008       2019-03-20   
4     236552     Pietra Aragão               8008       2022-11-06   

                  Curso        Área de Conhecimento Data de Conclusão  \
0              Botânica         Ciências Biológicas        2015-10-14   
1  Engenharia Sanitária                 Engenharias        2019-03-03   
2              Farmácia           Ciências da Saúde        2009-10-08   
3            Fisiologia         Ciências Biológicas        2017-12-07   
4           Comunicação  Ciências Sociais Aplicadas        2014-07-01   

                                         Instituição  
0  URI - Universidade Regional Integrada do Alto ...  
1                  FAVAP - Faculdades Vale do Piancó  
2     UNIVILLE - Universidade da Região de Joinville  
3             CUSC - Centro Universitário São Camilo  
4          Unifametro - Centro Universitário Fametro

#### Renomeia o dataframe

In [10]:
df_func = df_1
df_func

matricula                  Nome  local de trabalho Data de Ingresso  \
0        236548      Sra. Laura Silva               8008       2021-06-03   
1        236549        Henrique Ramos               8008       2020-06-28   
2        236550          Marina Lopes               8008       2015-12-08   
3        236551       Dra. Laura Dias               8008       2019-03-20   
4        236552         Pietra Aragão               8008       2022-11-06   
...         ...                   ...                ...              ...   
5995     242543         Renan Almeida               9909       2023-07-03   
5996     242544     Stephany da Cunha               9909       2015-10-01   
5997     242545         Clarice Gomes               9909       2015-02-20   
5998     242546  Ana Vitória Oliveira               9909       2022-01-30   
5999     242547        Fernanda Ramos               9909       2016-09-06   

                                           Curso         Área de Conhecimento  \
0                                       Botânica          Ciências Biológicas   
1                           Engenharia Sanitária                  Engenharias   
2                                       Farmácia            Ciências da Saúde   
3                                     Fisiologia          Ciências Biológicas   
4                                    Comunicação   Ciências Sociais Aplicadas   
...                                          ...                          ...   
5995  Recursos Florestais e Engenharia Florestal            Ciências Agrárias   
5996                                    História             Ciências Humanas   
5997                                  Morfologia          Ciências Biológicas   
5998                                   Filosofia             Ciências Humanas   
5999                                      Letras  Lingüística, Letras e Artes   

     Data de Conclusão                                        Instituição  
0           2015-10-14  URI - Universidade Regional Integrada do Alto ...  
1           2019-03-03                  FAVAP - Faculdades Vale do Piancó  
2           2009-10-08     UNIVILLE - Universidade da Região de Joinville  
3           2017-12-07             CUSC - Centro Universitário São Camilo  
4           2014-07-01          Unifametro - Centro Universitário Fametro  
...                ...                                                ...  
5995        2011-09-16                 UNESA - Universidade Estácio de Sá  
5996        2013-06-22    CESMAC - Centro de Estudos Superiores de Maceió  
5997        2011-01-01  USCS - Universidade Municipal de São Caetano d...  
5998        2019-05-02  FACELI - Faculdade de Ensino Superior de Linhares  
5999        2011-02-18  CEETEPS - Centro Estadual de Educação Tecnológ...  

[6000 rows x 8 columns]

#### exploração básica

##### Mostrar informação sobre os campos do dataframe

In [11]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6000 entries, 0 to 5999
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   matricula             6000 non-null   int64 
 1   Nome                  6000 non-null   object
 2   local de trabalho     6000 non-null   int64 
 3   Data de Ingresso      6000 non-null   object
 4   Curso                 6000 non-null   object
 5   Área de Conhecimento  6000 non-null   object
 6   Data de Conclusão     6000 non-null   object
 7   Instituição           6000 non-null   object
dtypes: int64(2), object(6)
memory usage: 375.1+ KB


### Dados de Dependência

#### Carrega o primeiro arquivo com dados das dependências

In [12]:
#no labblite

df_2 = pd.read_csv('local_part_01.csv', sep=";")
df_2.head(2)

COD NOME_LOCAL  COD_LOCAL NIVEL
0  7840        RIO          1     S
1  7850     MANAUS          2     S

##### Renomeia o dataframe

In [13]:
df_dep01 = df_2
df_dep01.head(2)

COD NOME_LOCAL  COD_LOCAL NIVEL
0  7840        RIO          1     S
1  7850     MANAUS          2     S

#### Carrega o primeiro arquivo com dados das dependências

In [14]:
# no labblite
df_3 = pd.read_csv('local_part_02.csv', sep=";")
df_3.head(2)


COD          NOME_LOCAL  COD_LOCAL NIVEL
0  229945  ASS.LEGISLATIVA-BA          6     D
1  230070       MEX-7.R C MEC         35     D

##### Renomeia o dataframe

In [15]:
df_dep02 = df_3
df_dep02.head(2)

COD          NOME_LOCAL  COD_LOCAL NIVEL
0  229945  ASS.LEGISLATIVA-BA          6     D
1  230070       MEX-7.R C MEC         35     D

#### exploração básica

##### Mostrar informação sobre os campos do dataframe

In [16]:
df_dep01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21501 entries, 0 to 21500
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COD         21501 non-null  int64 
 1   NOME_LOCAL  21501 non-null  object
 2   COD_LOCAL   21501 non-null  int64 
 3   NIVEL       21501 non-null  object
dtypes: int64(2), object(2)
memory usage: 672.0+ KB


In [17]:
df_dep02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13379 entries, 0 to 13378
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COD         13379 non-null  int64 
 1   NOME_LOCAL  13379 non-null  object
 2   COD_LOCAL   13379 non-null  int64 
 3   NIVEL       13379 non-null  object
dtypes: int64(2), object(2)
memory usage: 418.2+ KB


#### União dos dois dataframes

In [18]:
lista_dep = [df_dep01, df_dep02]

df_dep = pd.concat(lista_dep)

##### Mostrar registros carregados

In [19]:
df_dep.head(5000)

COD           NOME_LOCAL  COD_LOCAL NIVEL
0      7840                  RIO          1     S
1      7850               MANAUS          2     S
2      7855    PRESIDENTE VARGAS          3     S
3      7860               SANTOS          4     S
4      7870    CAMPOS GOYTACAZES          5     S
...     ...                  ...        ...   ...
4995  53760          PSO MARILIA       7839     S
4996  53765             PSO MAUA       7840     S
4997  53780          PSO TAUBATE       7843     S
4998  53785   PSO PAE S.PAULO II       7844     S
4999  53790  PSO PAE S.PAULO III       7845     S

[5000 rows x 4 columns]

##### Mostrar informação sobre os campos do dataframe

In [20]:
df_dep.info()

<class 'pandas.core.frame.DataFrame'>
Index: 34880 entries, 0 to 13378
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COD         34880 non-null  int64 
 1   NOME_LOCAL  34880 non-null  object
 2   COD_LOCAL   34880 non-null  int64 
 3   NIVEL       34880 non-null  object
dtypes: int64(2), object(2)
memory usage: 1.3+ MB


#### União os dois dataframes de Dependências

- os valores do dataframe "df_dep02" vão ser acrescentado no final do dataframe "df_dep01"

In [21]:
# cria lista contendo os dois dataframe
lista_df_dep = [df_dep01, df_dep02]

# faz a união utilizando o método do objeto pd
df_dep = pd.concat(lista_df_dep)

# imprime o dataframe resultante da união dos dois dataframes
df_dep

COD            NOME_LOCAL  COD_LOCAL NIVEL
0        7840                   RIO          1     S
1        7850                MANAUS          2     S
2        7855     PRESIDENTE VARGAS          3     S
3        7860                SANTOS          4     S
4        7870     CAMPOS GOYTACAZES          5     S
...       ...                   ...        ...   ...
13374  449475  DIOPE/ATAC COMEX MPE       9600     P
13375  449480  CENOP N PF/CONSORCIO       3901     P
13376  449490             USI/GESAE       8896     P
13377  449495       DITEC/GECAP6/G6       9909     P
13378  449500     DITEC/PROC. DE TI       9903     P

[34880 rows x 4 columns]

## Transformação dos dados

### Combinação dos dataframes de funcionários e dependências

Referências:
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html
- https://medium.com/data-hackers/pandas-combinando-data-frames-com-merge-e-concat-10e7d07ca5ec
- https://www.w3schools.com/python/pandas/ref_df_merge.asp
- https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html
- https://medium.com/@aneuk3/sql-joins-defcf817e8cf

In [22]:
df_func_dep_juncao = df_func.merge(df_dep,
                                   how="left",
                                   left_on="local de trabalho",
                                   right_on="COD_LOCAL"
                                  )

### Análise do resultado

In [23]:
df_func_dep_juncao

matricula              Nome  local de trabalho Data de Ingresso  \
0          236548  Sra. Laura Silva               8008       2021-06-03   
1          236548  Sra. Laura Silva               8008       2021-06-03   
2          236548  Sra. Laura Silva               8008       2021-06-03   
3          236548  Sra. Laura Silva               8008       2021-06-03   
4          236548  Sra. Laura Silva               8008       2021-06-03   
...           ...               ...                ...              ...   
211423     242547    Fernanda Ramos               9909       2016-09-06   
211424     242547    Fernanda Ramos               9909       2016-09-06   
211425     242547    Fernanda Ramos               9909       2016-09-06   
211426     242547    Fernanda Ramos               9909       2016-09-06   
211427     242547    Fernanda Ramos               9909       2016-09-06   

           Curso         Área de Conhecimento Data de Conclusão  \
0       Botânica          Ciências Biológicas        2015-10-14   
1       Botânica          Ciências Biológicas        2015-10-14   
2       Botânica          Ciências Biológicas        2015-10-14   
3       Botânica          Ciências Biológicas        2015-10-14   
4       Botânica          Ciências Biológicas        2015-10-14   
...          ...                          ...               ...   
211423    Letras  Lingüística, Letras e Artes        2011-02-18   
211424    Letras  Lingüística, Letras e Artes        2011-02-18   
211425    Letras  Lingüística, Letras e Artes        2011-02-18   
211426    Letras  Lingüística, Letras e Artes        2011-02-18   
211427    Letras  Lingüística, Letras e Artes        2011-02-18   

                                              Instituição     COD  \
0       URI - Universidade Regional Integrada do Alto ...   55825   
1       URI - Universidade Regional Integrada do Alto ...  265140   
2       URI - Universidade Regional Integrada do Alto ...  302470   
3       URI - Universidade Regional Integrada do Alto ...  311400   
4       URI - Universidade Regional Integrada do Alto ...  318955   
...                                                   ...     ...   
211423  CEETEPS - Centro Estadual de Educação Tecnológ...  448080   
211424  CEETEPS - Centro Estadual de Educação Tecnológ...  448390   
211425  CEETEPS - Centro Estadual de Educação Tecnológ...  448695   
211426  CEETEPS - Centro Estadual de Educação Tecnológ...  449055   
211427  CEETEPS - Centro Estadual de Educação Tecnológ...  449495   

                  NOME_LOCAL  COD_LOCAL NIVEL  
0         DIREC-CLI. VAR. PF       8008     S  
1       DIREC/RELACIONAMENTO       8008     K  
2       DIREC/SOL BENEFICIOS       8008     P  
3         DIREC/GEREN VAREJO       8008     P  
4        DIREC/ESTR CLIENTES       8008     P  
...                      ...        ...   ...  
211423    DITEC/UECII/GEPLAT       9909     P  
211424      DITEC/GECAP12/G4       9909     P  
211425    DITEC/UECII/GECAP6       9909     P  
211426   DITEC/UECII/GECAP12       9909     P  
211427       DITEC/GECAP6/G6       9909     P  

[211428 rows x 12 columns]

RESULTADO DA ANÁLISE:


- Reparou-se que os registros de funcionários duplicaram, isso ocorreu pelo fato do valor atribuído à coluna COD_LOCAL (do df_dep) estar vinculado a mais de um cadastro de dependência. Nota-se no print que o código 8008 está vinculado a diversas UORs:
    - 55825 DIREC-CLI. VAR. PF,
    - 265140 DIREC/RELACIONAMENT
    - 302470 DIREC/SOL BENEFICI
    - 311400 DIREC/GEREN VAR
    - 318955 DIREC/ESTR CLIE

- Com o objetivo de se manter a coesão dos dados, será aplicado um filtro para manter apenas registros que contenham o valor S na coluna NIVEL


### Exclusão de duplicação tutilizando a filtragem de apenas registros com a coluna "NIVEL" contendo valor 'S' (que representa o nível mais alto de uma cadeia de dependências e suas subordinadas)

In [24]:
df_func_dep_juncao = df_func_dep_juncao[df_func_dep_juncao["NIVEL"] == "S"]

In [25]:
# imprime os valores do dataframe após a exclusão dos valores duplicados
df_func_dep_juncao

matricula                  Nome  local de trabalho Data de Ingresso  \
0          236548      Sra. Laura Silva               8008       2021-06-03   
32         236549        Henrique Ramos               8008       2020-06-28   
64         236550          Marina Lopes               8008       2015-12-08   
96         236551       Dra. Laura Dias               8008       2019-03-20   
128        236552         Pietra Aragão               8008       2022-11-06   
...           ...                   ...                ...              ...   
210993     242543         Renan Almeida               9909       2023-07-03   
211080     242544     Stephany da Cunha               9909       2015-10-01   
211167     242545         Clarice Gomes               9909       2015-02-20   
211254     242546  Ana Vitória Oliveira               9909       2022-01-30   
211341     242547        Fernanda Ramos               9909       2016-09-06   

                                             Curso  \
0                                         Botânica   
32                            Engenharia Sanitária   
64                                        Farmácia   
96                                      Fisiologia   
128                                    Comunicação   
...                                            ...   
210993  Recursos Florestais e Engenharia Florestal   
211080                                    História   
211167                                  Morfologia   
211254                                   Filosofia   
211341                                      Letras   

               Área de Conhecimento Data de Conclusão  \
0               Ciências Biológicas        2015-10-14   
32                      Engenharias        2019-03-03   
64                Ciências da Saúde        2009-10-08   
96              Ciências Biológicas        2017-12-07   
128      Ciências Sociais Aplicadas        2014-07-01   
...                             ...               ...   
210993            Ciências Agrárias        2011-09-16   
211080             Ciências Humanas        2013-06-22   
211167          Ciências Biológicas        2011-01-01   
211254             Ciências Humanas        2019-05-02   
211341  Lingüística, Letras e Artes        2011-02-18   

                                              Instituição    COD  \
0       URI - Universidade Regional Integrada do Alto ...  55825   
32                      FAVAP - Faculdades Vale do Piancó  55825   
64         UNIVILLE - Universidade da Região de Joinville  55825   
96                 CUSC - Centro Universitário São Camilo  55825   
128             Unifametro - Centro Universitário Fametro  55825   
...                                                   ...    ...   
210993                 UNESA - Universidade Estácio de Sá  40165   
211080    CESMAC - Centro de Estudos Superiores de Maceió  40165   
211167  USCS - Universidade Municipal de São Caetano d...  40165   
211254  FACELI - Faculdade de Ensino Superior de Linhares  40165   
211341  CEETEPS - Centro Estadual de Educação Tecnológ...  40165   

                NOME_LOCAL  COD_LOCAL NIVEL  
0       DIREC-CLI. VAR. PF       8008     S  
32      DIREC-CLI. VAR. PF       8008     S  
64      DIREC-CLI. VAR. PF       8008     S  
96      DIREC-CLI. VAR. PF       8008     S  
128     DIREC-CLI. VAR. PF       8008     S  
...                    ...        ...   ...  
210993        DITEC/UEC II       9909     S  
211080        DITEC/UEC II       9909     S  
211167        DITEC/UEC II       9909     S  
211254        DITEC/UEC II       9909     S  
211341        DITEC/UEC II       9909     S  

[4433 rows x 12 columns]

### Redução ou Seleção de colunas

In [26]:
df_func_dep = df_func_dep_juncao[["matricula","Nome","Data de Ingresso","Área de Conhecimento","Data de Conclusão","COD","NOME_LOCAL"]]

In [27]:
df_func_dep.head(2)

matricula              Nome Data de Ingresso Área de Conhecimento  \
0      236548  Sra. Laura Silva       2021-06-03  Ciências Biológicas   
32     236549    Henrique Ramos       2020-06-28          Engenharias   

   Data de Conclusão    COD          NOME_LOCAL  
0         2015-10-14  55825  DIREC-CLI. VAR. PF  
32        2019-03-03  55825  DIREC-CLI. VAR. PF

### renomeando colunas

In [28]:
df_func_dep.columns = ["cd_func", "nm_func", "dt_ingso", "area_cmnto", "dt_conc", "cd_uor", "nm_uor_rdz"]

In [29]:
df_func_dep.head(2)

cd_func           nm_func    dt_ingso           area_cmnto     dt_conc  \
0    236548  Sra. Laura Silva  2021-06-03  Ciências Biológicas  2015-10-14   
32   236549    Henrique Ramos  2020-06-28          Engenharias  2019-03-03   

    cd_uor          nm_uor_rdz  
0    55825  DIREC-CLI. VAR. PF  
32   55825  DIREC-CLI. VAR. PF


## Gravação de Dados

In [30]:
# no labblite
df_func_dep.to_csv("func_dep.csv", sep=';',
                   index=False, 
                   encoding='utf-8')